### PR TITLE
gh-101277: Isolate itertools, part 3/4

### PR DESCRIPTION
**Pull request by erlend-aasland**
Add accumulate, compress, filterfalse, count, pairwise, zip_longest, and product types to module state

* Issue: gh-101277

### CODE REVIEW

File: Modules/clinic/itertoolsmodule.c.h [hunk 1]
```diff
@@ -102,7 +102,7 @@ static PyObject *
 pairwise_new(PyTypeObject *type, PyObject *args, PyObject *kwargs)
 {
     PyObject *return_value = NULL;
-    PyTypeObject *base_tp = &pairwise_type;
+    PyTypeObject *base_tp = clinic_state()->pairwise_type;
```
Comment (erlend-aasland) [inline Modules/clinic/itertoolsmodule.c.h:105]: I really like how more compact the clinic changes are, post #101411 🪄

**Comment by erlend-aasland** [comment]
@kumaraditya303: we're failing `test_finalize_structseq` of `test_embed`. the type version contract is invalidated[^1] and the `Py_TPFLAGS_VALID_VERSION_TAG` assert in `_PyType_Lookup` is triggered here:

https://github.com/python/cpython/blob/c4de6b1d52304a0a9cdfafc1dad5098993710404/Objects/typeobject.c#L4172-L4178

AFAICS, the contract between `tp_version_tag` and `Py_TPFLAGS_VALID_VERSION_TAG` seems to hold all across the APIs[^2] implemented typeobject.c, and from what I can see, the runtime uses only these APIs to mess with the type version.

Any ideas?

[^1]: `tp_version_tag` is set, but `tp_flags & Py_TPFLAGS_VALID_VERSION_TAG` is false
[^2]: `PyType_Modified`, `assign_version_tag`, etc.

**Comment by kumaraditya303** [comment]
Hmm, debugging shows that its is `sys.int_info` for which it is failing, this PR has nothing to with that so I guess it is a runtime bug somewhere. 

Note for self:
```console

Loop #2 ---\nobject address  : 0x5633734aa7e0\nobject refcount : 30\nobject type     : 0x563373352ce0\nobject type name: type\nobject repr     : <class 'sys.int_info'>\n_testembed: Objects/typeobject.c:4176: _PyType_Lookup: Assertion `_PyType_HasFeature(type, Py_TPFLAGS_VALID_VERSION_TAG)'
```

**Comment by erlend-aasland** [comment]
> Hmm, debugging shows that it is `sys.int_info` for which it is failing, this PR has nothing to with that so I guess it is a runtime bug somewhere.

Hm, but it does not fail on `main`; it only fails with this patch applied. I've tried (locally) to pull in latest `main`, just to be sure.

**Comment by erlend-aasland** [comment]
It's not just `int_info`. If you try to comment out `int_info` from the attribute list in `_test_embed_structseq.py`, it still fails. Commenting out these, and these only, attributes makes the test succeed:

<details>

```diff
diff --git a/Lib/test/_test_embed_structseq.py b/Lib/test/_test_embed_structseq.py
index 868f9f83e8..c5e4cd9746 100644
--- a/Lib/test/_test_embed_structseq.py
+++ b/Lib/test/_test_embed_structseq.py
@@ -27,10 +27,10 @@ def test_sys_attrs(self):
         for attr_name in (
             'flags',          # FlagsType
             'float_info',     # FloatInfoType
-            'hash_info',      # Hash_InfoType
-            'int_info',       # Int_InfoType
-            'thread_info',    # ThreadInfoType
-            'version_info',   # VersionInfoType
+            #'hash_info',      # Hash_InfoType
+            #'int_info',       # Int_InfoType
+            #'thread_info',    # ThreadInfoType
+            #'version_info',   # VersionInfoType
         ):
             with self.subTest(attr=attr_name):
                 attr = getattr(sys, attr_name)
```

</details>

**Comment by erlend-aasland** [comment]
Reverting ecb447f629ba78df2abdcb27697850d9e9679a08 fixes the problem. I suspect the problem lies within the ~220 lines of code implementing that type.

**Comment by erlend-aasland** [comment]
FTR, Kumar found the culprit. See #101696.

**Review by kumaraditya303 — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks, Kumar! Let's land.

**Pull request merged**